### ISSUE TITLE
Dramatically slow rendering (scrolling) since Ubuntu yakkety, xorg eats 23% cpu capacity.

### ISSUE THREAD

**Issue by CKD4XXH** [rdesktop/rdesktop#91]
Similar to issue #27. I connect from a (X)ubuntu to a (virtual) WinXP in a local network, via RDP, with rdesktop. Start, for example notepad, open a 3MiB text file (longest line 100 chars, ascii, crlf), and hit the PageDown key 18 times. The scrolling starts, and will go on autonomously (the keys have remembered). In linux I see that Xorg eats 23% of cpu. The "top" (in the default, "Irix" mode) shows 89%, for Xorg. My CPU is Intel Core i5 M 430, it has 2 cores and 4 "threads". I do not know whether the slow rendering in linux can cause (via flowcontrol and lossless (lapse-less) characteristics) slowing back the windows program. (Or, the windows program already finished, but I see the "past"?..)
I do not know that "what could mean" this slowing. Does it mean (usually), for example, that some hardware acceleration does not work? Or could it mean that instead of text, bitmap is transfered? (I use "rdesktop -a 16 -z -P .." so use compression.) Or some "frames" which could be skipped, are not skipped?..

Basically my wish to rdesktop testers would be merely to put into the "list of tests to do" the above test (some windows, notepad xxtext, pgdn, watch performance degradation).
Maybe not (or not only) into rdesktop's testlist, but into the Ubuntu release's testlist. Or Xorg's test list, or.. (help and forwarding appreciated).

The performance drop makes the remote windows machine unusable. When I started on the windows an Innosetup package-making, my whole linux became unresponsive. I shot down the RDP connection - and reconnected later (when the package-making already finished).

I experience this performance drop only in Yakkety Ubuntu, and only in "real" linux. Under Virtualbox there is no drop. There is no drop in Xenial Ubuntu.

VGA driver is i915.

**Comment by gohrner** [comment]
> Under Virtualbox there is no drop.

I'd say that would also indicate a video driver issue and no direct relation to X11.

**Comment by CKD4XXH** [comment]
I have switched to Remmina, with which I do not notice serious slowing-downs.

**Comment by kobaz** [comment]
Definitely has to do with something like the method of rendering.  Something is not being hardware accelerated.  I'm using a Radeon HD 7750 and remmina works fine, and so does krdc.

But xfreerdp and rdesktop have a major cpu hit and massive lag when drawing (which makes sense due to the fork history since they are tightly related)

**Comment by uglym8** [comment]
I'm working on adding shared memory support for graphic ops among the other things.
So it should make it faster a little bit.

Regarding CPU utilization:
Also we could employ some SIMD instructions for image handling (e.g. converting from RGB565 to ARGB32 etc). 
I've done it as a hobby project but never have completed it to the extents it could possibly be merged with upstream.

Regarding massive lag when using rdesktop/xfreerdp: 
I didn't experienced any lags for drawing ops when using rdesktop. 
Maybe we have different usage patterns.
Could you verify that you're having the same behaviour (i.e. lags) when using another distribution/version?

**Comment by gohrner** [comment]
I switched to Remmina due to the performance issues, so I cannot really test any more. Actually I forgot about the concrete circumstances why rdesktop became unusuable for me.

**Comment by kobaz** [comment]
In regards to "Maybe we have different usage patterns."

My usage pattern is:
Connect to plain vanilla windows 7, windows 10, or windows 20XX server.

Results: Excessively high cpu load when moving/resizing windows, switching windows to full screen, and doing any other activity that causes a lot of graphical activity like redrawing.

This has occurred after a hardware change.  I previously never had any issues with using nvidia graphics cards. And upon switching to AMD Radeon, I started having these performance problems.  But the problems only occur with rdesktop/xfreerdp.  I do not have any lag issues with remmina or with any other app, 2d or 3d or otherwise.

This leads me to believe that something is not being 2d accelerated in rdesktop/xfreerdp when using AMD video cards.

**Comment by gohrner** [comment]
It wasn't as bad for me, but I'm using Intel graphic cards.

If I remember correctly, I was often using IBM Data Studio, an Eclipse based SQL/DB2 frontend, in the remote session, and whenever it started to log something into its SQL console, things became unusable.

rdesktop CPU usage would jump to 100%, the whole session becoming unresponsive and starting to lag while rdesktop is trying to keep up with the repaints triggered by Data Studio. I think the Windows console window triggered similar/the same behaviour. While text was scrolling, the RDP session was not usable any more.

For me it suddenly started after some Ubuntu upgrade, it was fine before and no hardware change was involved.

I have no performance problems (at least which I know of) with any other software, including Remmina, to which I switched back then.

**Comment by lhzw** [comment]
xfreerdp handles this problem a little better, but its keyboard mapping is terrible.
rdesktop just hangs when I scroll a webpage and an image comes out.
